{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyOECe4GM5goaWPlRbNoT55y",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/luuloi/GWAS_Introduction_2023/blob/main/gwas_quantitative_traits.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Prepare session by install necessary packages"
      ],
      "metadata": {
        "id": "7SKgzQhF06FY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install rpy2\n",
        "!pip install -q condacolab"
      ],
      "metadata": {
        "id": "A5HGA8Fb0ylO",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 571
        },
        "outputId": "83101656-6aec-4503-95ea-fc6209116a3f"
      },
      "execution_count": 26,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting rpy2\n",
            "  Downloading rpy2-3.5.13.tar.gz (218 kB)\n",
            "\u001b[?25l     \u001b[90mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[32m0.0/218.8 kB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K     \u001b[90mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[32m218.8/218.8 kB\u001b[0m \u001b[31m9.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
            "  Preparing metadata (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting jinja2\n",
            "  Downloading Jinja2-3.1.2-py3-none-any.whl (133 kB)\n",
            "\u001b[2K     \u001b[90mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[32m133.1/133.1 kB\u001b[0m \u001b[31m14.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting tzlocal\n",
            "  Downloading tzlocal-5.0.1-py3-none-any.whl (20 kB)\n",
            "Requirement already satisfied: cffi>=1.10.0 in /usr/local/lib/python3.10/site-packages (from rpy2) (1.15.1)\n",
            "Requirement already satisfied: pycparser in /usr/local/lib/python3.10/site-packages (from cffi>=1.10.0->rpy2) (2.21)\n",
            "Collecting MarkupSafe>=2.0\n",
            "  Downloading MarkupSafe-2.1.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (25 kB)\n",
            "Building wheels for collected packages: rpy2\n",
            "  Building wheel for rpy2 (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for rpy2: filename=rpy2-3.5.13-cp310-cp310-linux_x86_64.whl size=262728 sha256=604b7a6fb4c9c2c27a22f38a29a4f0187eb3442f5cc520fac7ff88eb2bcffcd7\n",
            "  Stored in directory: /root/.cache/pip/wheels/b3/10/14/333faaf40f9cd461b2cd5a785a32d593b88e08b1834f081bd5\n",
            "Successfully built rpy2\n",
            "Installing collected packages: tzlocal, MarkupSafe, jinja2, rpy2\n",
            "Successfully installed MarkupSafe-2.1.3 jinja2-3.1.2 rpy2-3.5.13 tzlocal-5.0.1\n",
            "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\u001b[33m\n",
            "\u001b[0m"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.colab-display-data+json": {
              "pip_warning": {
                "packages": [
                  "jinja2",
                  "markupsafe",
                  "tzlocal"
                ]
              }
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\u001b[33m\n",
            "\u001b[0m"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Ignore rpy2's warnings\n",
        "\n",
        "import warnings\n",
        "from rpy2.rinterface import RRuntimeWarning\n",
        "\n",
        "warnings.filterwarnings(\"ignore\", category=RRuntimeWarning)"
      ],
      "metadata": {
        "id": "scD2EeD9_fme"
      },
      "execution_count": 27,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Initialize conda\n",
        "\n",
        "import condacolab\n",
        "condacolab.install()"
      ],
      "metadata": {
        "id": "a3kXfJv70-5r",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "736b82cb-548e-4a61-b25b-ae12245802a1"
      },
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "âœ¨ðŸ°âœ¨ Everything looks OK!\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Install plink, bedtools\n",
        "\n",
        "!conda install -c bioconda plink\n",
        "!conda install -c bioconda bedtools"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "s3IDHAqL3QaM",
        "outputId": "a674c41f-450c-47ad-a81d-7bc8fd0ed3a1"
      },
      "execution_count": 43,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "Solving environment: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 23.1.0\n",
            "  latest version: 23.7.2\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c conda-forge conda\n",
            "\n",
            "Or to minimize the number of packages updated during conda update use\n",
            "\n",
            "     conda install conda=23.7.2\n",
            "\n",
            "\n",
            "\n",
            "# All requested packages already installed.\n",
            "\n",
            "Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Solving environment: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 23.1.0\n",
            "  latest version: 23.7.2\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c conda-forge conda\n",
            "\n",
            "Or to minimize the number of packages updated during conda update use\n",
            "\n",
            "     conda install conda=23.7.2\n",
            "\n",
            "\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - bedtools\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    bedtools-2.31.0            |       hf5e1c6e_2         1.5 MB  bioconda\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:         1.5 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  bedtools           bioconda/linux-64::bedtools-2.31.0-hf5e1c6e_2 \n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages\n",
            "                                                                        \n",
            "Preparing transaction: \\ \b\bdone\n",
            "Verifying transaction: / \b\bdone\n",
            "Executing transaction: \\ \b\bdone\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Set up R environment by installing required packages."
      ],
      "metadata": {
        "id": "WiOPlMmtFe_w"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Load `rpy2`"
      ],
      "metadata": {
        "id": "U-QQIimc1IYS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%load_ext rpy2.ipython"
      ],
      "metadata": {
        "id": "9Bj1-Nl11GCP",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "94128a09-17a4-4d50-d9d9-411a147bf63d"
      },
      "execution_count": 30,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "The rpy2.ipython extension is already loaded. To reload it, use:\n",
            "  %reload_ext rpy2.ipython\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "`pacman` will make your life easier and I have include `BiocManager` for bio-oriented packages."
      ],
      "metadata": {
        "id": "aRwL8PiRFlp0"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 31,
      "metadata": {
        "id": "HDe1tVAzrtTK",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "5eae5b42-527c-4203-be12-6167c2943271"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "R[write to console]: Loading required package: pacman\n",
            "\n"
          ]
        }
      ],
      "source": [
        "%%R\n",
        "if (!require(\"pacman\")) install.packages(\"pacman\")\n",
        "library(\"pacman\")\n",
        "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "\n",
        "# Function to check and install missing packages\n",
        "check_and_install_packages <- function(packages) {\n",
        "    for (package in packages) {\n",
        "        if (!require(package, character.only = TRUE)) {\n",
        "            if (package %in% rownames(available.packages(repos = BiocManager::repositories())))\n",
        "                BiocManager::install(package, character.only = TRUE)\n",
        "        else\n",
        "            install.packages(package, character.only = TRUE)\n",
        "        require(package, character.only = TRUE)\n",
        "        }\n",
        "    }\n",
        "}"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "We will use the following packages for today lecture.\n",
        "\n",
        "* [qqman](https://cran.r-project.org/web/packages/qqman/vignettes/qqman.html)\n",
        "* [SKAT](https://cran.r-project.org/web/packages/SKAT/index.html)\n",
        "* [GWASTools](https://bioconductor.org/packages/release/bioc/manuals/GWASTools/man/GWASTools.pdf)\n",
        "* [gdsfmt](https://bioconductor.org/packages/release/bioc/manuals/gdsfmt/man/gdsfmt.pdf)\n",
        "* [SNPRelate](https://www.bioconductor.org/packages/release/bioc/manuals/SNPRelate/man/SNPRelate.pdf)\n",
        "* [GENESIS](https://bioconductor.org/packages/release/bioc/manuals/GENESIS/man/GENESIS.pdf)"
      ],
      "metadata": {
        "id": "E3_1ARqvF706"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "check_and_install_packages(c(\"qqman\", \"SKAT\", \"GWASTools\", \"SNPRelate\", \"GENESIS\"))"
      ],
      "metadata": {
        "id": "g5b_SFymr4eQ",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "764417f9-c380-42c0-d4c4-03b513e8c949"
      },
      "execution_count": 32,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "R[write to console]: Loading required package: qqman\n",
            "\n",
            "R[write to console]: \n",
            "\n",
            "R[write to console]: For example usage please run: vignette('qqman')\n",
            "\n",
            "R[write to console]: \n",
            "\n",
            "R[write to console]: Citation appreciated but not required:\n",
            "\n",
            "R[write to console]: Turner, (2018). qqman: an R package for visualizing GWAS results using Q-Q and manhattan plots. Journal of Open Source Software, 3(25), 731, https://doi.org/10.21105/joss.00731.\n",
            "\n",
            "R[write to console]: \n",
            "\n",
            "R[write to console]: Loading required package: SKAT\n",
            "\n",
            "R[write to console]: Loading required package: Matrix\n",
            "\n",
            "R[write to console]: Loading required package: SPAtest\n",
            "\n",
            "R[write to console]: Loading required package: RSpectra\n",
            "\n",
            "R[write to console]: Loading required package: GWASTools\n",
            "\n",
            "R[write to console]: Loading required package: Biobase\n",
            "\n",
            "R[write to console]: Loading required package: BiocGenerics\n",
            "\n",
            "R[write to console]: \n",
            "Attaching package: â€˜BiocGenericsâ€™\n",
            "\n",
            "\n",
            "R[write to console]: The following objects are masked from â€˜package:statsâ€™:\n",
            "\n",
            "    IQR, mad, sd, var, xtabs\n",
            "\n",
            "\n",
            "R[write to console]: The following objects are masked from â€˜package:baseâ€™:\n",
            "\n",
            "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
            "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
            "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
            "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
            "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
            "    table, tapply, union, unique, unsplit, which.max, which.min\n",
            "\n",
            "\n",
            "R[write to console]: Welcome to Bioconductor\n",
            "\n",
            "    Vignettes contain introductory material; view with\n",
            "    'browseVignettes()'. To cite Bioconductor, see\n",
            "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
            "\n",
            "\n",
            "R[write to console]: Loading required package: SNPRelate\n",
            "\n",
            "R[write to console]: Loading required package: gdsfmt\n",
            "\n",
            "R[write to console]: SNPRelate -- supported by Streaming SIMD Extensions 2 (SSE2)\n",
            "\n",
            "R[write to console]: Loading required package: GENESIS\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Load data"
      ],
      "metadata": {
        "id": "ytv3R8Ofr8yS"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Read directly from URL"
      ],
      "metadata": {
        "id": "8m-nQCJrWnrn"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "# Function to read and display data\n",
        "read_and_display <- function(file_path, col_names=NULL, sep=\" \", header=FALSE) {\n",
        "  data <- read.table(file=file_path, sep=sep, header=header, na.strings=\"NA\")\n",
        "  if (!is.null(col_names)) {\n",
        "    names(data) <- col_names\n",
        "  }\n",
        "\n",
        "  cat(\"\\n--- Head of the Data ---\\n\")\n",
        "  print(head(data))\n",
        "  cat(\"\\n--- Dimensions of the Data ---\\n\")\n",
        "  print(dim(data))\n",
        "\n",
        "  return(data)\n",
        "}"
      ],
      "metadata": {
        "id": "9rs3NlpQ6mpP"
      },
      "execution_count": 33,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "# Read and display Transferrin individuals' info\n",
        "fam_info <- read_and_display(\"https://faculty.washington.edu/tathornt/sisg/Transferrin.fam\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "IVpj2s5A9GlK",
        "outputId": "e3bd0e48-eff6-4b63-ef96-e20ed454d8c6"
      },
      "execution_count": 34,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "--- Head of the Data ---\n",
            "    V1  V2  V3    V4 V5 V6\n",
            "1  355 883 681 10680  2 -9\n",
            "2  355 884 681 10680  1 -9\n",
            "3 3004 885 682 10681  2 -9\n",
            "4 3155 886 683 10682  1 -9\n",
            "5 1629 887 684 10683  2 -9\n",
            "6 1747 888 685 10684  2 -9\n",
            "\n",
            "--- Dimensions of the Data ---\n",
            "[1] 4861    6\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "# Read and display SNP info\n",
        "snp_info <- read_and_display(\"https://faculty.washington.edu/tathornt/sisg/Transferrin.bim\", sep=\"\\t\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4kWmhh5a9HRJ",
        "outputId": "669bfc21-51f8-4006-fb67-a5240d5679dc"
      },
      "execution_count": 35,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "--- Head of the Data ---\n",
            "  V1         V2 V3      V4 V5 V6\n",
            "1  1  rs3934834  0  995669  T  C\n",
            "2  1  rs3737728  0 1011278  A  G\n",
            "3  1  rs6687776  0 1020428  T  C\n",
            "4  1  rs9651273  0 1021403  A  G\n",
            "5  1  rs4970405  0 1038818  G  A\n",
            "6  1 rs12726255  0 1039813  G  A\n",
            "\n",
            "--- Dimensions of the Data ---\n",
            "[1] 281313      6\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "# Read and display Transferrin phenotype info\n",
        "tpheno_info <- read_and_display(\"https://faculty.washington.edu/tathornt/sisg/Tr.pheno\", col_names=c(\"FAMID\", \"ID\", \"Transferrin\"))\n",
        "cat(\"\\n--- Summary of Transferrin ---\\n\")\n",
        "print(summary(tpheno_info[[\"Transferrin\"]]))\n",
        "cat(\"\\n--- NA counts in Transferrin ---\\n\")\n",
        "print(table(is.na(tpheno_info[[\"Transferrin\"]])))\n",
        "hist(tpheno_info[[\"Transferrin\"]], xlab=\"Transferrin\", main=\"Histogram of Transferrin\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 878
        },
        "id": "gMwtRHo09IoS",
        "outputId": "3e1053aa-d3da-4914-f843-8fe616290917"
      },
      "execution_count": 36,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "--- Head of the Data ---\n",
            "  FAMID  ID Transferrin\n",
            "1   355 883      -0.815\n",
            "2   355 884          NA\n",
            "3  3004 885          NA\n",
            "4  3155 886      -0.299\n",
            "5  1629 887          NA\n",
            "6  1747 888       1.182\n",
            "\n",
            "--- Dimensions of the Data ---\n",
            "[1] 4861    3\n",
            "\n",
            "--- Summary of Transferrin ---\n",
            "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's \n",
            "-3.4960 -0.6195 -0.0450  0.0335  0.5818  4.5150    2499 \n",
            "\n",
            "--- NA counts in Transferrin ---\n",
            "\n",
            "FALSE  TRUE \n",
            " 2362  2499 \n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAMAAABKCk6nAAAC4lBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFTU1NUVFRVVVVXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlra2tsbGxtbW1vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6goKChoaGioqKjo6OkpKSlpaWnp6epqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzOzs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7///90dUCGAAARbElEQVR4nO3deXwU5R3H8QciUogkHIlBAgnhihoiBJFT5Ugi1oOAQAVEqxUvrIoX8WpRBFFARKu0URAEbVXQUEHrgQhCBESMaKkmEo4IKGhCruf/zszObpKdZ9idnZnsPL/9fl4vNjr7PLPP7tscIwsP44h0LNoLQO4GYOIBmHgAJh6AiQdg4gGYeAAmHoCJB2DiAZh4ACYegIkHYOIBmHgAJh6AiQdg4gGYeAAmHoCJB2DiAZh4UgKzOMdO9cPFvxvpyHgH1+Ro8gBXMValfJjGFvHCRxoPv80q7Jz1Hpa/Sv2YyXxNC3O8oaZr8lJSAjftJnvA09ka7ePdU6aks5wpU54Lc7w0SQmsfjn8bGzHDqO38Szls+5+Xn3vOW16PlzL+TcXtR30Hsvmp1hGUcIrfEd+p84Tf+A1rOdbafG3Hh7Xfvj/fCfzTxilTM/SH0D7T8cwcdN58eMqAg/nG39wWmq73L36WH2GsqbG0R5KVuCqxNTnlnbreOLNs9ni7Xw8u2zBIHYbb+jLJszNYIOU55WQ+efNvyW1WVbIxij/Fn/50ng2YMHlbKrvZP4JxcPZ7W/qD+D72hA8ceRLw9lNgYfTxtfntJ7/clLvet9Y/21c42gvJRWwLw14P7vwCN+36xTvrXyJ/oL1qOMHzjiz6lPWo54vZRcqz4tt4vzEB1s5b9emVvm3PXwGu45/xXpp5wpMUFTX+h9ABw6a+Dnfzs5tfDh1/AdsVFXVQ2yDb6z/VgXWR3spqYBHjBw5MtkH3DCCtc6681uuAa9gBcqAdLZnFbua8xIfsPrpvjgrIZ6xE5y1quMPsnn8Z5aknSswQQQcPPEn1rXx4dTxf/P9h7ZYH6vfqsD6aC8lFXDT78HVRZNTWKcyHXi8ck8P9vVKFfgzDVi9anmTXbB1X3vVSfm3QvYUP8G6aOcKTBAAGyYeYymND6eOX8HGbFEq0y+NGm8Doz2UrMA/f1zGG2axVxXgcr6Lpdbx8rizaj5kafX82QDwA2wOL2UC4MAEE+BmE1WywMOp47eqX4W/+ugYgB2tOfBmNvgfRTlxu/hgdmMxn8DGL8phj/O6NFbwWEYA+AXW/9XB/dii34KBAxNMgJtNVMkCD6eObxjAZizs1rEcwI4W9CV6dU77DkOUH4BfT2pXyKvuSWnT+3nl3p1ZZw5dzYbqr3r1lLPSVryTmHTYAByYIAZuNlEj8z+cNv7Q9K4JY7ZzALd4DWVb1W+gU6O9Dg9FDLgPu+mlTFYc7XV4KFrA/JurOscPXB3tVXgpYsAoOAATD8DEAzDxAEw8ABMPwMQDMPEATDwAEw/AxAMw8QBMPAATD8DEAzDxAEw8ABMPwMQDMPEATDwAEw/AxAMw8QBMPAATD8DEAzDxAEw8ABMPwMQDMPEATDwywL/Mf1LQimgvK+qRAd41drmg4dFeVtSjAzx1t6DR0V5W1AMw8QBMPAATD8DEAzDxAEw8ABMvHODq0pJ9ta6vxGYAFhca+MA1bVMyktvfcLQFVmMjAIsLDZz3wHHl9tBtBe4vxk4AFhcauFOd9uFUF7eXYi8AiwsNfO5H2oeN2W4vxV4AFhca+N0uubPuu31s8sYWWI2NACwujJ+ij7/2+Jwn3vjF/bXYCsDicJlEPFwmES+Sy6Rfv9P6psrFdVkOwOIiuUxaN1lrxINuLSqSACzOxmXSa8ucX07kAVicjcskAMuQjcskAMtQmL9dWH7EeAzAMhQaeDLne7POjBv+ffAdAJah0MBtOR9dWFs158rgOwAsQ2EBJ1dzXpsUfAeAZSi8z+BvOP+2R/AdAJah0MAsqfc5M/mX3ecH3wFgGQoNXFXx9Wd7+I/GTZcBLEM23lUJYBkCMPEATDwAEw/AxCMOPHS7oF0N0V5sC0YcuM8kQT33R3uxLRhx4ItEByd+F+3FtmAAJh6AiQdg4gGYeAAmHoCJB2DiAZh4ACYegIkHYOIBmHgAJh6AiQdg4gGYeAAmHoCJB2DiAZh4ACYegIkHYOIBmHgAJh6AiQdg4gGYeAAmHoCJB2Di2diUI2rA5WsELQSwMBubckQN+KmCu4xdDWBhNvYujBrwgmUCtUUAFmZj70IAy5CMm3IA2EIybsoBYAvJuCkHgC0k42USgC0k42USgC2EyyTiRXKZVDxTK+9e11Z1+gBsoUguk076dh99dolrqzp9ALYQLpOIh8sk4oX724XrjIcALEOhgfdqJe3dG3wHgGUojI2xUrKUzsjKCr4DwDIUGviDgY9UcZ5ivAPAMhTG9+DqR/tvBLCshfVDVunY6Ya9RwEsR2H+FP3yBOMxAMuQjO+qBLCFAEw8ABMPwMQDMPEATDwAEw/AxAMw8QBMPAATD8DEAzDxAEw8ABMPwMQDMPEATDwAEw/AxAMw8QBMPAATD8DEAzDxAEw8ABMPwMQDMPEATDwAEw/AxAMw8QBMPAATD8DEAzDxAEw8ABMPwMTDrivEw64rxNOBB8wzfdL0dl3plTPMWN+iKD0bl9OB192YfNHC74Uj6G1OmV4sOLjQU9tdO1fgS3Tdh7O6jlj2s3EEvc0pYxP4eNHvE66+Jf1dwwh6u67EIvDaie1Gv6T8GLUnzTiE3K4rsQg8+Okfff/wgGAMtcukWASueqKa//BktWgEvcukWASedvWv/OikGaIRxsukE/rWdotbYoGCAGwhHThVvRaqPUc0wniZtEHfnPI+11cnDsAW0oHT1Gvg3d1FI3CZJHU68MqkguuvSFgjGoHLJKnzXwf/98X5RT+Kh+AySebC/O3ChooGwzEAy5AO/M6wtFQl0YjSvLQ7K7Nbp3wSfAeAZUgHPnftF+qzFI0Yfe+W6y95pvblIcF3AFiGdOB88xGJdfwoq+INnYLvALAM6cCFm0xHZHzJ+WzO9/UKvgPAMuT/Df/WiWbfg1d22aDcbur2TPAdAJYhHbjE9yyFQ77erw4oNhwHsAwFLpNqDlidCmAZ0oEPTWjTkd+x2dJUAMuQDjxm4Yl0vm2wpakAliEdWPkROZ3zPpamAliGdOALdirAe/tbmgpgGfK/bbZzfnxB8npLUwEsQ/6fostfnL/ioLWpAJahWPzDZ7EInKnVz9JUAMuQDrxF6d83P21pKoBlqNmX6CssTQWwDDUFPpllaSqAZajJ9+C+8XdbmgpgGWryPXhbmbWpAJYhXCbFBnBab3/hTwWwDOnAC//w7ufrx88rKSkJfyqAZUgHzq5XbmrOszQVwDKkA3dX/z90eVdLUwEsQzrw3E4Tri/o+KilqQCWIf9P0XteeOL5ndamAliG8Ka72ADGm+6IA+NNd8SB8aY74sB40x1xYLzpjjgw3nRHHDg3gqkAliEd+No19ZanAliGdOCc+PbdTP58sGkAliEdeIfvWVqaCmAZ0oDVN9tNP2Z1KoBlSAOOU36lVFidCmAZAjCAzQKwDPmAy8rKkncoN5amAliGNGDmz9JUAMuQZlrrz9JUAMtQLO5dCOBmYVMOqQsNTG/vQgA3C3sXSl1oYGzKIXWhgbEph9SF8VM0NuWQOVwmxTwwLpOkzuOXSV9tFHQzgMMvksuktyZrDZvt2qoC5d55l7EMAIdfJJdJp45qFS11bVWBcncKLC4FcPh5/DIJwHbz+GUSgO0W/m8BVwYfALAMhQbeO6b7teq+pG2D7wCwDIUGHvnwjvm99wFY0kIDd6jnvDizAsByFho4U91XdlX2fgBLWWjgfyasVW+7twq+A8AyFMZP0WXl6u1PRcHHASxDHv/bZgFsNwAD2CwAyxCAAWwWgGUIwAA2C8AyBGAAmwVgGQIwgM2iBTyn4ElB1v72Ti8GYL3rpi039ux495+iywHYD1woOLgFwC4HYLsBGMBmAViGAAxgswAsQwAGsFkAliEAA9gsAMsQgAFsFoBlCMAANgvAMgRgAJsFYBkCMIDNArAMARjAZgFYhgAMYLMALEMABrBZAJYhAAPYLADLEIABbBaAZcjju64A2G4e33UFwHbz+K4rALabxzenBLDdPL45JYDthl1XYh44RnZdiWFgXCbJHC6TYh7YeJl0dLvWvMUurksPwHaL5DLpvZlaeQ+4tqpAALYbLpNOB/yftGGCern/vJ0Ll0mnA940RnhS95+3c+EyCcB6B4MPAFiGwgcmvvNZ7AIf0QMwUeBWcb4MIwEsQ6GBZz/m+4jPYKLAtQUl2kcAEwX2R3yLdwAbArAMARjAZgFYhgAMYLMALEMABrBZAJYhAAPYLADLkGeAG2bNFNQDwDbzDHDNoGJBHQBsM+8ADxe9lgkAthmA9QBsCMAyBGA9ABsCsAwBWA/AhgAsQwDWA7AhAMsQgPUAbAjAMgRgPQAbArAMAVgPwIYALEMA1rMAfOFRUU6+GA4GYD0LwAmXCerzkZOvhnMBWM8CcAfRwdvfd/LVcC4A6wHYEICbBuAQAdidAKwHYEMAbhqAQwRgdwKwHoANAbhpAA4RgN0JwHoANgTgpl2cfr6xjLlOvkQRFQ3gmu8ElUoOPOTvgoOrZoqe6qHIX3PrRQN4/QWTjE2kCPx4puCZThoa+WtuvWgAvz1b8Fp8ThF47o2ioaMjf82tB2A9ABsCcNMA3BiAWzDn9i48IvqB8XvR+WIdOCVP0LjdotevLhLTZjm3d+FA0TuVzv7T/cauiXHgQZsFdRslePmyX2kB4HC3eB8tei59/7Lc2JUxDixcVLf3BQcfumejoFJngY17F741WWvE/c3GDRJ9BncYKTjYO1NwMO9s0fwzRQeTRQ+VkyYa2kd0MP4SwcH08wQHR6WEvagugwUH+/cUDU0XHWwv+gzuwURNcxbYuHdhje+NwJWnmo37VfRm4UOigz8dER09HP7Bn8J+qPBPWilclM2Tio8KTyo8WFkpOlrtLLDp3oVIhmzsXYhkyMZ1MJIhABMPwMQDMPEATDwAEw/AxAMw8QBMPM8A1/cT/S6p3dKlOWm/g+68rp4Brsl346yuvHnClZNO/dGNswLYMycFcEQBGMAeOSmAIwrAAPbISckD117uxlnzpDnp9Ao3zuohYO7KW0Zi/KTcS8DIlQBMPAATD8DEAzDxAEw8ABMPwMTzDvBb5yaOsvQHI0NX3L/zOMf/B5EL61T7gO11/qTcQ8BliZ/UF45x9JTHkz6tLZzk6CldWada9cAU6sBrON+R6ugp14xTkNta+sOWoXNhnWqPPpJFHFht/hRHT/f4LOUmxYUvpw6vU6m0f1UMAG/IKHP0fHPuU24yShw9p5rT61Qau5FTBl7SpcsSzldl7nP2tE/crtwkO3xSF9bJ+cvTOGngkxUVJ/mb2U7/wPvGpZyXtxf8BVC2cn6dnE9ISkk5I2md4+flHgFWO9p9v9On/CXp/drbZjh8UhfW6YvyZ7BaUau2SpWOnvO98ztf5ewZ3VmnFnVg5E4AJh6AiQdg4gGYeAAmHoCJB2DiAZh4ACYegIkHYOIBmHgAJh6AiQdg4gGYeAAmHoCJB2Di0Qe+LC6OxcXFHTe7/9bO/2p+4PPebi+pJaMPrJS65TR3dg7+sy21R9xcS0sXM8A7cqbn8eV9e476gZcMmJOfuYGfmt6r59TfJrTOXLcuu1fuYW2E+kv5DNZHUChmgHfHr+aH2+7nN93Cd7d+j68dztfmN9Tf8wlPLCvvvJMvnKCNUH8pwPoICsUOcLt6zn/jfFU+353I+Rdp/OPU9VXKnYllL+ZzfqJNjTpC/aUC+0ZQKHaAu3He8NeLhvTN5btTufZrzaiEP55UgOfHp6endzygjlB/qcD6CALFDrDi9Xr2cf5KbhO+ytwFCvBK3+bm/sMAli4/8NKr+LH8oX6+JY82NNzwlAJ88OxSvu0OAMubH/jw0My8T1Nm63yHr+zRc/IJBZivz+5z4ccARjIGYOIBmHgAJh6AiQdg4gGYeAAmHoCJB2DiAZh4ACYegIkHYOIBmHgAJh6AiQdg4gGYeP8H98vRRi+hgJcAAAAASUVORK5CYII=\n"
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "# Read and display height phenotype info\n",
        "hpheno_info <- read_and_display(\"https://faculty.washington.edu/tathornt/sisg/Ht.pheno\", col_names=c(\"FAMID\", \"ID\", \"Height\"))\n",
        "cat(\"\\n--- Summary of Height ---\\n\")\n",
        "print(summary(hpheno_info[[\"Height\"]]))\n",
        "cat(\"\\n--- NA counts in Height ---\\n\")\n",
        "print(table(is.na(hpheno_info[[\"Height\"]])))\n",
        "hist(hpheno_info[[\"Height\"]], xlab=\"Height\", main=\"Histogram of Height\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 878
        },
        "id": "PU7Dtars9KQR",
        "outputId": "ff7e6b6d-0466-4ba1-d483-8de8241f3171"
      },
      "execution_count": 39,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "--- Head of the Data ---\n",
            "  FAMID  ID      Height\n",
            "1   355 883          NA\n",
            "2   355 884 -1.01219122\n",
            "3  3004 885 -1.11122366\n",
            "4  3155 886          NA\n",
            "5  1629 887 -0.04663134\n",
            "6  1747 888  1.59969343\n",
            "\n",
            "--- Dimensions of the Data ---\n",
            "[1] 4861    3\n",
            "\n",
            "--- Summary of Height ---\n",
            "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's \n",
            "-4.0274 -0.7340 -0.0299 -0.0204  0.6232  4.5077    2025 \n",
            "\n",
            "--- NA counts in Height ---\n",
            "\n",
            "FALSE  TRUE \n",
            " 2836  2025 \n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAHgCAMAAABKCk6nAAAC6FBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFTU1NUVFRVVVVXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBiYmJjY2NlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5udnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWnp6epqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////nY40YAAARkElEQVR4nO3de3gU1RnH8YOIKUSuSRoRSMAgIOEiiIJQK+Ei1laSilZFLNJq1bZWWyxRsRUUC2qLqNjGK4q11XoBRYuXWoEqEQERaWtrAlGCiEabQEjOv53ZnV2TOXuc2cmZZN93ft/nMYHZmbMn8xGyB5YZIRHrREdPAIUbgJkHYOYBmHkAZh6AmQdg5gGYeQBmHoCZB2DmAZh5AGYegJkHYOYBmHkAZh6AmQdg5gGYeQBmHoCZB2DmsQAWnY0N9cE3vjbJz7MYfMpwowtcL0S99elCcbssv+HLzX8RNW0Z9WoxfZVr9EQtn+VL4DY+X+ixAG7ZJW074bPF6q8aPVkSuI3PF3osgO2zvWFKr+6TN8liIcR82fDzvl0GXt8o5XsnZ419XoyUB8Wgih4PyM3Te/f57gfykBj4ZEH2pXtndDv1P/HBEgecbh1e7Br9wwv7dZ26I26aGE+KrE2jjp5R4zxfBscFuL5nvzuXH9ur7omvizvekDPFGbeOFZfJ5uNF2U2DxFjr6+wx9Kev/S+3y4pyUWL9LPvM5dli9K1nigvigyUOWHOquPyJ1qM3jTliyf25RU32s7QYr9uJPxgg5sr482VwpIHjxYDfFyfVyl1vH5RF1m+Zb4kBh+WeI4+q/7sY0CSXi5Osr1O8IGXd+o1Sdu3SaP1su5wjLpLviONiYyUPsEQfc42+XpxeX3+dWGs/S8vxnpMviaEy9nyZHGngiZMmTcqLAzdPFEcU/+Sf8RP+oCi1digU21eJs6WsjIPYvyDvKO6RLUSdFJ0Oy1+Km+WnIjc2VvKAlsDO6PfEoe+wn6XFeJ0OylpxDIDDy/U9uKHi3HzRu8oBnmk9MkC8+5ANsiEGYr8oekKM2rirmw1s/axc/EbWiZzYWMkDWgI7oz8oSl63qrIPaj3eJyIfwOHVGvjTV6tk85XiYeuEV8u3Rb/Dsrrz0YdeFgVN8ndJ4GvFArlTpABOHpACeKMYJuU7r3xiH9R6PAe4uqPOgK+4AL8mxv2hYkznt+U4MXeNLBMzbx8jFsnDBaL0xkFJ4LvFiIfHDRG3/88NnDwgBXDzaDFn6bG9qu0hWo8XA449XwbHBVg+MqZb91OsF8B/zO1aLuuvzu9SdJf16Jbio8Y/IsY7wA3nHV3w4DM9c/cqwMkDVGD50exjepS8ER+i1Xgx4NjzZXB0gf3UXLXR/s57QaaO1w4xBx4sLrl3qDD2e6jp8doh3sDyve/0yT7xkcwdL/yYAyMAMw/AzAMw8wDMPAAzD8DMAzDzAMw8ADMPwMwDMPMAzDwAMw/AzAMw8wDMPAAzD8DMAzDzAMw8ADMPwMwDMPMAzDwAMw/AzAMw8wDMPAAzD8DMAzDzAMy8SAK/eYva7o6eVEhFEvjGa37vbtbTHT2pkIom8B+2uvsZgBkFYOYBmHkAZh6AmQdg5gGYeQBmHoCZB2DmAZh5AGYegJkHYOYBmHkAZh6AmQdg5gGYeQBmHoCZB2DmAZh5AGYegJkH4NY17Kzc1Rj6TMLq8H6lXwK4RXvOycoflNft+/vbYTZhdP2ICe56ALhF0649YH386LLS8CcTSvNXKZonA7hFvQ/HPh3MCXsqIQVgj4a9Evu0bmTYUwkpAHv0bM7UK39x+ZS8de0wmzACsFcHHl20YPHjn4U/l3ACsGe0l0kA9oj6MgnAHmGZRLogy6Q/Tos15obQZmUyAHukXSY9usL8dEIIwB5pl0kAplAblkkAppDPvy6srlW3AZhC3sDnSrmj+KjOp/7X/QCAKeQNnCXl5PLG+gXfdj8AYAr5As5rkLIx1/0AgCnk71fwe1L+c4D7AQBTyBtY5Bb1/YHc1n+J+wEAU8gbuL7m3Q3b5e41ygMAplAb3lUJYAoBGMC6AEwhAANYF4ApBGAA6wIwhQAMYF0AphCAAawLwBQCMIB1AZhCAAawLgBTCMAA1gVgCgEYwLoATCEAA1gXgCkEYADrAjCFAAxgXZkI/Nm/lC4FcMAyEfjHZ8xy1wfAActE4Cv+rNCNAnDAAEwhAANYF4ApBGAA6wIwhQAMYF0AphCAAawLwBRqwxXfAUyhNlzxHcAUasMV3wFMoTbcGAvAFGJ2xXcAu2N2xXcAu2N2xXcAu8MyKfLAWCaRLsgy6WD8PtoVy0OcV8AA7C7IMunJc2NNuCa0WQUOwO6wTIo8cDSWSZctWufuxbqO/mpMhGVSvNPPmOtu4uqO/mpM5PevC59SN7EC/uadyqbyRzv6qzGRN/COWLk7drgfADCFfNx1Jb/Y6sjiYvcDAKaQN/D6E2+olzJffQDAFPLxPbhh4Yh1AKaarxdZO6fMVm5sB2Aa+XwVfX+Zug3AFMK7KgGsC8AUAjCAdQGYQgAGsC4AUwjAANYFYAoBGMC6AEwhAANYF4ApBGAA6wIwhQAMYF0AphCAAawLwBQCMIB1AZhCAAawLgBTCMAA1gVgCgEYwLoATCEARwN49M3/SvtQAFPIAX5qbt7JS/+b3qEAplDyt+jDL195zMQVn6ZxKIAplAQ+UPGtHmf/sPBZ/4cCmEIO8GPf7Tr53v1Sbi/wfyiAKeQAj1u2O/6Da/0fCmAKOcD1ixvkB7c0pHUogCnkAF949hdy/6w5aR0KYAo5wP3sS8o29k3rUABTyAEusNfAW/un3icKV3znDvxQbunFZ/VIefXNaFzxnTuw/PfKJRW7U+4RjRtjsQfWF40bY3EHfmZCQT+rVHtE44rv3IGHPfaW/TWl2iMaV3znDjz9K3aJxBXfuQOXv/AV+2CZRLjEX/gf0VP3PRjLJNI5wJXxrynVHuoyaU/8riTXLW2PCaYXgN0ll0mH9mj2UJdJr8yPVXZdqDMLFIDdOcAflXXpJa94LdUeWCaRzgEuWVpXKDeNS7VH5i6Tfjx4uLuuAHblAB8nZaGUg1PukrHLpLlrFJQhAHblAI/aYgHvGKHdrbmmWdkGYAol3jbbZ3p2aV7KGzTunFbwk30jj8j/m/sBAFMo8Sq6euWSBz9Mucfkn79+8Wm3Nd5/ivsBAFPI+2+Teh6W+0W9bO7tfgDAFHKAh8YakmqPQdukvEbKXce5HwAwhRzg162em7cs1R4P5ay1Pr5w7G3uBwBMoVa/RZ+Vcpd337c+VK5RtgOYQi2BP1duMPqVAZhCLb4HH5/9s7QOBTCFWnwP3lSV3qEAphDlf+EPYB8l3vhelMj/oQCmkAO89HvP/uPpmTdXVlb6PxTAFHKARzZZHw6dkNahAKaQA9zf/nPo6mPSOhTAFHKAb+pddnFpr4VpHQpgCiVeRW+/e/FdW9I7FMAU8n7TnTYAU8j7TXfaAEwh7zfdaQMwhXy86U4XgCnk8013qQIwhbzfdKcNwBTyftOdNgBTyAGeGuBQAFPIAT5/dVPahwKYQg7wmOxux2r+fbA2AFPIAd4c/5rSOhTAFIoB22+2m/1JuocCmEIx4M7Wf/k16R4KYAoBGMC6AEyhOHBVVVXeZutDWodyBz5nyDR349O4pUWGFAMWidI6lDvwmUuVTYvu6+AvOf1ipo2J0joUwBTCG98BrAvAFAIwgHUBmEIABrAuAFMIwADO2AuCA9hH3sCZe0FwAPvIGzhz75sEYB9Rvm8SgH3kDZy5FwQHsI+8gTP3guAA9pGPV9FRvSB4ZICxTOINjGUSc2Ask5gDq8ukt1fGmvfr0GblLwD7KMgyaefqWFctDm1W/gKwj7BMijwwlkncgZ32uTcAmELewDtK+p+/2/qc5X4AwBTyBp50/eYlRbsAzBa4e5OUa4bWAJgr8FD7nnarRr4PYKbAf+rxmP2xfyf3AwCmkI9X0VXV9sePK9zbAUwhvKsSwLoATCEAA1gXgCkEYADrAjCFAAxgXQCmEIABrAvAFAIwgHUBmEIABrAuAFMIwADWBWAKARjAugBMIQADWBeAKQRgAOsCMIUADGBdAKYQgAGsC8AUAjCAdbUv8LbVSlMB7B0Z4PN+eJW73u0OPG+4cre7ac+351lIPzrALymnu7Ddgc+5Sdm07M72PAvpB2AA6wIwgA0G4GABGMC6AAxggwE4WGSu+A7gYJG54juAg0Xmiu8ADhaZG2MBOFhkbowF4GCRueI7gINF5orvAA4WlkmRB8YyiTkwlknMgdVl0kvzY5UuCG1WKQJwsIIsk2rfiHXzstBmlSIABwvLpMgDY5nEHdjpQ/cGAPMC7uC7rgA4WN7AtU4AZgrcqXM8ZU8A8wC+5sb4Z/wKZgrcWFoZ+wxgpsCJOvj2sgAOFt5VCWBdAAawwQAcLAADWBeAAWwwAAcLwADWBWAAGwzAwQIwgHUBGMAGA3CwAAxgXQAGsMEAHCwAA1gXgAFsMAAHC8AA1gVgu3l9h7sruKc9T4xHAG4j8JwFyqYVt7bnifEIwPEArBYi8CPzlU4AcKAyE/jsBx51lwPgQGUo8AblrPUFcKAAHA/AagC2A3CQAGwsAMcDsBqA7QAcJAAbC8DxAKwGYDsABwnAxgJwPACrAdgOwEECsLEAHA/AagC2SwG8cNS5SmtDO1ceATieSeAfzV3jbsnC0M6VRwCOZxT4KmVTRZSBfzpNqQ+ATZUBwFO2KOcjB8CmyoAbYwE4zNr7xljPrVQaHk3gunvVU/FmgDPqkbkbY725TulhddPoZUpdIwC88iLlTNxZopyJS/sqr0ZK5qnncJdZYPXGWE/Gl3YT57fab5hQOlLd1PkMpaPUTVlTlU1dJyubsk9TNnWfqGzqNV7ZlDNO2ZQ3VtnUd5Syqf8IZdPAE5RNg4com05Qz4QYqOw1tkDZdEqKA2ebBVZvjHVof6x9B1vt98V+pY/UTXt9bWr/A/d+rGyq3ads2qdu+rhW3eTzKdVN/qbfYBZYe2MsRKE23BgLUagN62BEIQAzD8DMAzDzAMw8ADMPwMwDMPMAzLxQgcer79UIXFGJubGKJ5gba9xoc2OddpF5g1CBJxsc6zzlFvPBu/Gv5sZabfBKd9suNzdWIgC3MQAbCsDBAnAbA7ChABwsALexSANPMzjW+XvNjXXTy+bGevxuc2O9c6W5sRKFCmzyXSAmx/qiydxYh+rNjSXrDI7lhD/JYh6AmQdg5gGYeQBmHoCZB2DmAZh5IQOvFzsMjfTksJ6n7zQy0poRfWbUGBlJmpyWnbnTlSxc4IYT8w3NuKrn35rKS0yMdCD3743ls0yMJE1Oy87c6fqycIEX3lBsCni1lJv7mRhp9QwLOSutf4Kpz9y07Mydri8LFXjniHqTM15ynolRFtl/op9v7rdVQ9OSxk9XvFCBp6yTBme8dlCViWEW/ML6MKjSxFCxDE1Lmj5dTmEB/zYn57f3XyiNzNgeS8pVQ9O6NoW2xfZfuuaZGUuam5aUpk5X68IC/rym5vOy3Pz8I3OfMjKWfGKkoVe+j39TyupuKS4LFShj05LS1OlqXdjrYFP/S+7v/76ZgeRnuS82XjbH0GDmphWPzq/gRKZmXNEpy2qfiaGeH97nO0YGkkanFYseMOrgAMw8ADMPwMwDMPMAzDwAMw/AzAMw8wDMPAAzD8DMAzDzAMw8ADMPwMwDMPMAzDwAMw/AzIsUcKOw36N+e/IujP8oSvxoq/PPT+5r9zmFXaSBG2sTP3KAm/PbfU5hF1Hgp0YeN3Vv7FfwLYVjVhbKrQN/NbBwrSzrVPxBR0/ScNEEru6zRS4ts4G39dzTMKNIbs16QK6cKGuzOnqKxosYcF5+fn73UrlyupR1XQ5ZwHeVSfm4BdxdyrcKAEy8RrG5pqbm16VySXZhYWGvPRbw4kuk3FAU+x5s/Qdg2iV/i34o/jrLAl4+S8o/A5hJSeAPv75TbrrCBt6YW3vwrCTwgc6fd/QcTRdNYPn0yMEnvRp7FX11/wkrBieA5fReGzp6koaLFHCqmqRcP66jJxFiUQfee/S25ktCuA53xhR1YHnPwIKZtd67kS3ywNwDMPMAzDwAMw/AzAMw8wDMPAAzD8DMAzDzAMw8ADMPwMwDMPMAzDwAM+//tXbZyNYDQhUAAAAASUVORK5CYII=\n"
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "\n",
        "# Download required data files (Replace the LINK placeholders with actual links)\n",
        "wget https://faculty.washington.edu/tathornt/sisg/Transferrin.bim -O Transferrin.bim\n",
        "wget https://faculty.washington.edu/tathornt/sisg/Transferrin.fam -O Transferrin.fam\n",
        "wget https://faculty.washington.edu/tathornt/sisg/Tr.pheno -O Tr.pheno\n",
        "wget https://faculty.washington.edu/tathornt/sisg/Ht.pheno -O Ht.pheno\n",
        "\n",
        "# Convert bam to bed\n",
        "bedtools bamtobed -i Transferrin.fam > sample.bed\n",
        "\n",
        "###  ASSOCIATION ANALYSIS with Transferrin  Phenotype ###\n",
        "plink --bfile Transferrin --pheno Tr.pheno --maf 0.05 --geno 0.01 --hwe 0.001 --assoc --out GWAS_T_add\n",
        "\n",
        "###  ASSOCIATION ANALYSIS with Height Phenotype ###\n",
        "plink --bfile Transferrin --pheno Ht.pheno --maf 0.05 --geno 0.01 --hwe 0.001 --assoc --out GWAS_H_add\n",
        "\n",
        "### LD of top association SNPs for Transferrin and Height ###\n",
        "plink --bfile Transferrin --extract SNPlistTransferrin.txt --r2 --out LD_T_Subset\n",
        "plink --bfile Transferrin --extract SNPlistHeight.txt --r2 --out LD_H_Subset\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "yAr3dSrf-LKB",
        "outputId": "97979876-4d33-479e-889c-3ab551608769"
      },
      "execution_count": 42,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "Logging to GWAS_T_add.log.\n",
            "Options in effect:\n",
            "  --assoc\n",
            "  --bfile Transferrin\n",
            "  --geno 0.01\n",
            "  --hwe 0.001\n",
            "  --maf 0.05\n",
            "  --out GWAS_T_add\n",
            "  --pheno Tr.pheno\n",
            "\n",
            "12982 MB RAM detected; reserving 6491 MB for main workspace.\n",
            "PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "Logging to GWAS_H_add.log.\n",
            "Options in effect:\n",
            "  --assoc\n",
            "  --bfile Transferrin\n",
            "  --geno 0.01\n",
            "  --hwe 0.001\n",
            "  --maf 0.05\n",
            "  --out GWAS_H_add\n",
            "  --pheno Ht.pheno\n",
            "\n",
            "12982 MB RAM detected; reserving 6491 MB for main workspace.\n",
            "PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "Logging to LD_T_Subset.log.\n",
            "Options in effect:\n",
            "  --bfile Transferrin\n",
            "  --extract SNPlistTransferrin.txt\n",
            "  --out LD_T_Subset\n",
            "  --r2\n",
            "\n",
            "12982 MB RAM detected; reserving 6491 MB for main workspace.\n",
            "PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/\n",
            "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
            "Logging to LD_H_Subset.log.\n",
            "Options in effect:\n",
            "  --bfile Transferrin\n",
            "  --extract SNPlistHeight.txt\n",
            "  --out LD_H_Subset\n",
            "  --r2\n",
            "\n",
            "12982 MB RAM detected; reserving 6491 MB for main workspace.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "--2023-08-14 03:14:25--  https://faculty.washington.edu/tathornt/sisg/Transferrin.bim\n",
            "Resolving faculty.washington.edu (faculty.washington.edu)... 128.208.60.35, 140.142.214.168\n",
            "Connecting to faculty.washington.edu (faculty.washington.edu)|128.208.60.35|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 \n",
            "Length: 7787053 (7.4M)\n",
            "Saving to: â€˜Transferrin.bimâ€™\n",
            "\n",
            "     0K .......... .......... .......... .......... ..........  0%  385K 20s\n",
            "    50K .......... .......... .......... .......... ..........  1%  800K 14s\n",
            "   100K .......... .......... .......... .......... ..........  1% 28.6M 10s\n",
            "   150K .......... .......... .......... .......... ..........  2% 1.79M 8s\n",
            "   200K .......... .......... .......... .......... ..........  3% 1.36M 8s\n",
            "   250K .......... .......... .......... .......... ..........  3% 54.2M 6s\n",
            "   300K .......... .......... .......... .......... ..........  4%  134M 5s\n",
            "   350K .......... .......... .......... .......... ..........  5% 1.80M 5s\n",
            "   400K .......... .......... .......... .......... ..........  5%  185M 5s\n",
            "   450K .......... .......... .......... .......... ..........  6% 1.40M 5s\n",
            "   500K .......... .......... .......... .......... ..........  7%  237M 4s\n",
            "   550K .......... .......... .......... .......... ..........  7% 89.7M 4s\n",
            "   600K .......... .......... .......... .......... ..........  8%  114M 3s\n",
            "   650K .......... .......... .......... .......... ..........  9%  196M 3s\n",
            "   700K .......... .......... .......... .......... ..........  9% 86.5M 3s\n",
            "   750K .......... .......... .......... .......... .......... 10%  213M 3s\n",
            "   800K .......... .......... .......... .......... .......... 11% 1.84M 3s\n",
            "   850K .......... .......... .......... .......... .......... 11%  248M 3s\n",
            "   900K .......... .......... .......... .......... .......... 12% 83.9M 2s\n",
            "   950K .......... .......... .......... .......... .......... 13% 1.44M 3s\n",
            "  1000K .......... .......... .......... .......... .......... 13%  146M 2s\n",
            "  1050K .......... .......... .......... .......... .......... 14%  202M 2s\n",
            "  1100K .......... .......... .......... .......... .......... 15%  180M 2s\n",
            "  1150K .......... .......... .......... .......... .......... 15%  124M 2s\n",
            "  1200K .......... .......... .......... .......... .......... 16%  181M 2s\n",
            "  1250K .......... .......... .......... .......... .......... 17%  120M 2s\n",
            "  1300K .......... .......... .......... .......... .......... 17%  177M 2s\n",
            "  1350K .......... .......... .......... .......... .......... 18%  197M 2s\n",
            "  1400K .......... .......... .......... .......... .......... 19%  255M 2s\n",
            "  1450K .......... .......... .......... .......... .......... 19%  208M 2s\n",
            "  1500K .......... .......... .......... .......... .......... 20%  224M 2s\n",
            "  1550K .......... .......... .......... .......... .......... 21%  199M 1s\n",
            "  1600K .......... .......... .......... .......... .......... 21%  223M 1s\n",
            "  1650K .......... .......... .......... .......... .......... 22% 1.93M 1s\n",
            "  1700K .......... .......... .......... .......... .......... 23%  132M 1s\n",
            "  1750K .......... .......... .......... .......... .......... 23%  156M 1s\n",
            "  1800K .......... .......... .......... .......... .......... 24%  243M 1s\n",
            "  1850K .......... .......... .......... .......... .......... 24%  183M 1s\n",
            "  1900K .......... .......... .......... .......... .......... 25%  265M 1s\n",
            "  1950K .......... .......... .......... .......... .......... 26% 1.45M 1s\n",
            "  2000K .......... .......... .......... .......... .......... 26%  141M 1s\n",
            "  2050K .......... .......... .......... .......... .......... 27%  177M 1s\n",
            "  2100K .......... .......... .......... .......... .......... 28%  308M 1s\n",
            "  2150K .......... .......... .......... .......... .......... 28%  119M 1s\n",
            "  2200K .......... .......... .......... .......... .......... 29%  320M 1s\n",
            "  2250K .......... .......... .......... .......... .......... 30%  156M 1s\n",
            "  2300K .......... .......... .......... .......... .......... 30% 64.1M 1s\n",
            "  2350K .......... .......... .......... .......... .......... 31% 92.5M 1s\n",
            "  2400K .......... .......... .......... .......... .......... 32%  236M 1s\n",
            "  2450K .......... .......... .......... .......... .......... 32%  323M 1s\n",
            "  2500K .......... .......... .......... .......... .......... 33%  335M 1s\n",
            "  2550K .......... .......... .......... .......... .......... 34%  283M 1s\n",
            "  2600K .......... .......... .......... .......... .......... 34%  342M 1s\n",
            "  2650K .......... .......... .......... .......... .......... 35%  309M 1s\n",
            "  2700K .......... .......... .......... .......... .......... 36%  103M 1s\n",
            "  2750K .......... .......... .......... .......... .......... 36% 40.9M 1s\n",
            "  2800K .......... .......... .......... .......... .......... 37% 57.2M 1s\n",
            "  2850K .......... .......... .......... .......... .......... 38% 61.3M 1s\n",
            "  2900K .......... .......... .......... .......... .......... 38% 64.4M 1s\n",
            "  2950K .......... .......... .......... .......... .......... 39%  172M 1s\n",
            "  3000K .......... .......... .......... .......... .......... 40%  370M 1s\n",
            "  3050K .......... .......... .......... .......... .......... 40%  363M 1s\n",
            "  3100K .......... .......... .......... .......... .......... 41%  342M 1s\n",
            "  3150K .......... .......... .......... .......... .......... 42%  266M 1s\n",
            "  3200K .......... .......... .......... .......... .......... 42%  215M 1s\n",
            "  3250K .......... .......... .......... .......... .......... 43%  270M 1s\n",
            "  3300K .......... .......... .......... .......... .......... 44%  289M 1s\n",
            "  3350K .......... .......... .......... .......... .......... 44%  247M 1s\n",
            "  3400K .......... .......... .......... .......... .......... 45% 2.45M 1s\n",
            "  3450K .......... .......... .......... .......... .......... 46% 59.8M 1s\n",
            "  3500K .......... .......... .......... .......... .......... 46% 55.8M 1s\n",
            "  3550K .......... .......... .......... .......... .......... 47% 56.2M 1s\n",
            "  3600K .......... .......... .......... .......... .......... 47% 53.6M 1s\n",
            "  3650K .......... .......... .......... .......... .......... 48% 61.6M 1s\n",
            "  3700K .......... .......... .......... .......... .......... 49% 63.3M 0s\n",
            "  3750K .......... .......... .......... .......... .......... 49% 69.2M 0s\n",
            "  3800K .......... .......... .......... .......... .......... 50%  355M 0s\n",
            "  3850K .......... .......... .......... .......... .......... 51%  360M 0s\n",
            "  3900K .......... .......... .......... .......... .......... 51%  378M 0s\n",
            "  3950K .......... .......... .......... .......... .......... 52% 1.65M 0s\n",
            "  4000K .......... .......... .......... .......... .......... 53% 42.1M 0s\n",
            "  4050K .......... .......... .......... .......... .......... 53%  328M 0s\n",
            "  4100K .......... .......... .......... .......... .......... 54%  344M 0s\n",
            "  4150K .......... .......... .......... .......... .......... 55%  328M 0s\n",
            "  4200K .......... .......... .......... .......... .......... 55%  311M 0s\n",
            "  4250K .......... .......... .......... .......... .......... 56%  289M 0s\n",
            "  4300K .......... .......... .......... .......... .......... 57%  350M 0s\n",
            "  4350K .......... .......... .......... .......... .......... 57%  295M 0s\n",
            "  4400K .......... .......... .......... .......... .......... 58%  374M 0s\n",
            "  4450K .......... .......... .......... .......... .......... 59%  352M 0s\n",
            "  4500K .......... .......... .......... .......... .......... 59%  348M 0s\n",
            "  4550K .......... .......... .......... .......... .......... 60%  168M 0s\n",
            "  4600K .......... .......... .......... .......... .......... 61%  283M 0s\n",
            "  4650K .......... .......... .......... .......... .......... 61%  345M 0s\n",
            "  4700K .......... .......... .......... .......... .......... 62%  324M 0s\n",
            "  4750K .......... .......... .......... .......... .......... 63%  288M 0s\n",
            "  4800K .......... .......... .......... .......... .......... 63%  349M 0s\n",
            "  4850K .......... .......... .......... .......... .......... 64%  307M 0s\n",
            "  4900K .......... .......... .......... .......... .......... 65%  335M 0s\n",
            "  4950K .......... .......... .......... .......... .......... 65%  256M 0s\n",
            "  5000K .......... .......... .......... .......... .......... 66%  353M 0s\n",
            "  5050K .......... .......... .......... .......... .......... 67%  345M 0s\n",
            "  5100K .......... .......... .......... .......... .......... 67%  355M 0s\n",
            "  5150K .......... .......... .......... .......... .......... 68%  225M 0s\n",
            "  5200K .......... .......... .......... .......... .......... 69%  246M 0s\n",
            "  5250K .......... .......... .......... .......... .......... 69%  353M 0s\n",
            "  5300K .......... .......... .......... .......... .......... 70%  321M 0s\n",
            "  5350K .......... .......... .......... .......... .......... 71%  311M 0s\n",
            "  5400K .......... .......... .......... .......... .......... 71%  332M 0s\n",
            "  5450K .......... .......... .......... .......... .......... 72%  351M 0s\n",
            "  5500K .......... .......... .......... .......... .......... 72% 67.1M 0s\n",
            "  5550K .......... .......... .......... .......... .......... 73% 43.5M 0s\n",
            "  5600K .......... .......... .......... .......... .......... 74% 63.4M 0s\n",
            "  5650K .......... .......... .......... .......... .......... 74% 63.2M 0s\n",
            "  5700K .......... .......... .......... .......... .......... 75% 66.0M 0s\n",
            "  5750K .......... .......... .......... .......... .......... 76% 54.8M 0s\n",
            "  5800K .......... .......... .......... .......... .......... 76% 65.0M 0s\n",
            "  5850K .......... .......... .......... .......... .......... 77% 64.7M 0s\n",
            "  5900K .......... .......... .......... .......... .......... 78% 64.5M 0s\n",
            "  5950K .......... .......... .......... .......... .......... 78% 56.6M 0s\n",
            "  6000K .......... .......... .......... .......... .......... 79% 66.0M 0s\n",
            "  6050K .......... .......... .......... .......... .......... 80% 65.7M 0s\n",
            "  6100K .......... .......... .......... .......... .......... 80% 63.4M 0s\n",
            "  6150K .......... .......... .......... .......... .......... 81% 58.1M 0s\n",
            "  6200K .......... .......... .......... .......... .......... 82% 65.0M 0s\n",
            "  6250K .......... .......... .......... .......... .......... 82% 64.1M 0s\n",
            "  6300K .......... .......... .......... .......... .......... 83% 5.32M 0s\n",
            "  6350K .......... .......... .......... .......... .......... 84% 55.6M 0s\n",
            "  6400K .......... .......... .......... .......... .......... 84% 59.8M 0s\n",
            "  6450K .......... .......... .......... .......... .......... 85% 58.0M 0s\n",
            "  6500K .......... .......... .......... .......... .......... 86%  123M 0s\n",
            "  6550K .......... .......... .......... .......... .......... 86%  305M 0s\n",
            "  6600K .......... .......... .......... .......... .......... 87% 58.2M 0s\n",
            "  6650K .......... .......... .......... .......... .......... 88% 90.1M 0s\n",
            "  6700K .......... .......... .......... .......... .......... 88% 1.61M 0s\n",
            "  6750K .......... .......... .......... .......... .......... 89% 37.5M 0s\n",
            "  6800K .......... .......... .......... .......... .......... 90%  202M 0s\n",
            "  6850K .......... .......... .......... .......... .......... 90%  298M 0s\n",
            "  6900K .......... .......... .......... .......... .......... 91%  352M 0s\n",
            "  6950K .......... .......... .......... .......... .......... 92%  322M 0s\n",
            "  7000K .......... .......... .......... .......... .......... 92%  375M 0s\n",
            "  7050K .......... .......... .......... .......... .......... 93%  347M 0s\n",
            "  7100K .......... .......... .......... .......... .......... 94%  323M 0s\n",
            "  7150K .......... .......... .......... .......... .......... 94%  289M 0s\n",
            "  7200K .......... .......... .......... .......... .......... 95%  333M 0s\n",
            "  7250K .......... .......... .......... .......... .......... 95%  345M 0s\n",
            "  7300K .......... .......... .......... .......... .......... 96%  350M 0s\n",
            "  7350K .......... .......... .......... .......... .......... 97%  318M 0s\n",
            "  7400K .......... .......... .......... .......... .......... 97%  346M 0s\n",
            "  7450K .......... .......... .......... .......... .......... 98% 72.5M 0s\n",
            "  7500K .......... .......... .......... .......... .......... 99% 64.2M 0s\n",
            "  7550K .......... .......... .......... .......... .......... 99% 55.2M 0s\n",
            "  7600K ....                                                  100% 8.46T=0.6s\n",
            "\n",
            "2023-08-14 03:14:26 (12.8 MB/s) - â€˜Transferrin.bimâ€™ saved [7787053/7787053]\n",
            "\n",
            "--2023-08-14 03:14:26--  https://faculty.washington.edu/tathornt/sisg/Transferrin.fam\n",
            "Resolving faculty.washington.edu (faculty.washington.edu)... 128.208.60.35, 140.142.214.168\n",
            "Connecting to faculty.washington.edu (faculty.washington.edu)|128.208.60.35|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 \n",
            "Length: 126193 (123K)\n",
            "Saving to: â€˜Transferrin.famâ€™\n",
            "\n",
            "     0K .......... .......... .......... .......... .......... 40%  401K 0s\n",
            "    50K .......... .......... .......... .......... .......... 81%  811K 0s\n",
            "   100K .......... .......... ...                             100% 40.1M=0.2s\n",
            "\n",
            "2023-08-14 03:14:27 (659 KB/s) - â€˜Transferrin.famâ€™ saved [126193/126193]\n",
            "\n",
            "--2023-08-14 03:14:27--  https://faculty.washington.edu/tathornt/sisg/Tr.pheno\n",
            "Resolving faculty.washington.edu (faculty.washington.edu)... 128.208.60.35, 140.142.214.168\n",
            "Connecting to faculty.washington.edu (faculty.washington.edu)|128.208.60.35|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 \n",
            "Length: 68827 (67K)\n",
            "Saving to: â€˜Tr.phenoâ€™\n",
            "\n",
            "     0K .......... .......... .......... .......... .......... 74%  403K 0s\n",
            "    50K .......... .......                                    100% 42.0M=0.1s\n",
            "\n",
            "2023-08-14 03:14:27 (541 KB/s) - â€˜Tr.phenoâ€™ saved [68827/68827]\n",
            "\n",
            "--2023-08-14 03:14:27--  https://faculty.washington.edu/tathornt/sisg/Ht.pheno\n",
            "Resolving faculty.washington.edu (faculty.washington.edu)... 128.208.60.35, 140.142.214.168\n",
            "Connecting to faculty.washington.edu (faculty.washington.edu)|128.208.60.35|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 \n",
            "Length: 103721 (101K)\n",
            "Saving to: â€˜Ht.phenoâ€™\n",
            "\n",
            "     0K .......... .......... .......... .......... .......... 49%  394K 0s\n",
            "    50K .......... .......... .......... .......... .......... 98% 1.45M 0s\n",
            "   100K .                                                     100% 2.40T=0.2s\n",
            "\n",
            "2023-08-14 03:14:28 (630 KB/s) - â€˜Ht.phenoâ€™ saved [103721/103721]\n",
            "\n",
            "Error: Failed to open Transferrin.bed.\n",
            "Error: Failed to open Transferrin.bed.\n",
            "Error: Failed to open Transferrin.bed.\n",
            "Error: Failed to open Transferrin.bed.\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "CalledProcessError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-42-0c326e0a0c11>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_cell_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'bash'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'\\n# Download required data files (Replace the LINK placeholders with actual links)\\nwget https://faculty.washington.edu/tathornt/sisg/Transferrin.bim -O Transferrin.bim\\nwget https://faculty.washington.edu/tathornt/sisg/Transferrin.fam -O Transferrin.fam\\nwget https://faculty.washington.edu/tathornt/sisg/Tr.pheno -O Tr.pheno\\nwget https://faculty.washington.edu/tathornt/sisg/Ht.pheno -O Ht.pheno\\n\\n###  ASSOCIATION ANALYSIS with Transferrin  Phenotype ###\\nplink --bfile Transferrin --pheno Tr.pheno --maf 0.05 --geno 0.01 --hwe 0.001 --assoc --out GWAS_T_add\\n\\n###  ASSOCIATION ANALYSIS with Height Phenotype ###\\nplink --bfile Transferrin --pheno Ht.pheno --maf 0.05 --geno 0.01 --hwe 0.001 --assoc --out GWAS_H_add\\n\\n### LD of top association SNPs for Transferrin and Height ###\\nplink --bfile Transferrin --extract SNPlistTransferrin.txt --r2 --out LD_T_Subset\\nplink --bfile Transferrin --extract SNPlistHeight.txt --r2 --out LD_H_Subset\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/_shell.py\u001b[0m in \u001b[0;36mrun_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m    332\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mline\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    333\u001b[0m       \u001b[0mcell\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m' '\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 334\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_cell_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmagic_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    335\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    336\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mrun_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2471\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2472\u001b[0m                 \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mmagic_arg_s\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2473\u001b[0;31m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2474\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2475\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/IPython/core/magics/script.py\u001b[0m in \u001b[0;36mnamed_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    140\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    141\u001b[0m                 \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mscript\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 142\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshebang\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    143\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    144\u001b[0m         \u001b[0;31m# write a basic docstring:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m<decorator-gen-103>\u001b[0m in \u001b[0;36mshebang\u001b[0;34m(self, line, cell)\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/IPython/core/magic.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(f, *a, **k)\u001b[0m\n\u001b[1;32m    185\u001b[0m     \u001b[0;31m# but it's overkill for just that one bit of state.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmagic_deco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 187\u001b[0;31m         \u001b[0mcall\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/IPython/core/magics/script.py\u001b[0m in \u001b[0;36mshebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    243\u001b[0m             \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflush\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    244\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraise_error\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m\u001b[0;34m!=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 245\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mCalledProcessError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutput\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    246\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    247\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_run_script\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mto_close\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mCalledProcessError\u001b[0m: Command 'b'\\n# Download required data files (Replace the LINK placeholders with actual links)\\nwget https://faculty.washington.edu/tathornt/sisg/Transferrin.bim -O Transferrin.bim\\nwget https://faculty.washington.edu/tathornt/sisg/Transferrin.fam -O Transferrin.fam\\nwget https://faculty.washington.edu/tathornt/sisg/Tr.pheno -O Tr.pheno\\nwget https://faculty.washington.edu/tathornt/sisg/Ht.pheno -O Ht.pheno\\n\\n###  ASSOCIATION ANALYSIS with Transferrin  Phenotype ###\\nplink --bfile Transferrin --pheno Tr.pheno --maf 0.05 --geno 0.01 --hwe 0.001 --assoc --out GWAS_T_add\\n\\n###  ASSOCIATION ANALYSIS with Height Phenotype ###\\nplink --bfile Transferrin --pheno Ht.pheno --maf 0.05 --geno 0.01 --hwe 0.001 --assoc --out GWAS_H_add\\n\\n### LD of top association SNPs for Transferrin and Height ###\\nplink --bfile Transferrin --extract SNPlistTransferrin.txt --r2 --out LD_T_Subset\\nplink --bfile Transferrin --extract SNPlistHeight.txt --r2 --out LD_H_Subset\\n'' returned non-zero exit status 2."
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "I9HDTaaxATpH"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}